File: src/gobby/code_index/languages.py
"""Language registry mapping languages to tree-sitter queries.

Each LanguageSpec defines file extensions, and tree-sitter query strings
for extracting symbols, imports, and calls from a given language.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, cast

logger = logging.getLogger(__name__)

# Cached parsers (tree-sitter Parser objects are reusable and thread-safe)
_parser_cache: dict[str, Any] = {}


@dataclass
class LanguageSpec:
    """Specification for a single language's tree-sitter queries."""

    extensions: list[str]
    symbol_query: str
    import_query: str = ""
    call_query: str = ""
    docstring_query: str = ""
    # Node types that indicate a class-like container (for parent linking)
    container_types: list[str] = field(default_factory=list)


# ── Query Definitions ────────────────────────────────────────────────────

_PYTHON_SPEC = LanguageSpec(
    extensions=[".py", ".pyi"],
    symbol_query="""
        (function_definition name: (identifier) @name) @definition.function
        (class_definition name: (identifier) @name) @definition.class
    """,
    import_query="""
        (import_statement) @import
        (import_from_statement) @import
    """,
    call_query="""
        (call function: (identifier) @name) @call
        (call function: (attribute attribute: (identifier) @name)) @call
    """,
    container_types=["class_definition"],
)

_JAVASCRIPT_SPEC = LanguageSpec(
    extensions=[".js", ".jsx", ".mjs", ".cjs"],
    symbol_query="""
        (function_declaration name: (identifier) @name) @definition.function
        (class_declaration name: (identifier) @name) @definition.class
        (method_definition name: (property_identifier) @name) @definition.method
        (export_statement declaration: (function_declaration name: (identifier) @name)) @definition.function
        (export_statement declaration: (class_declaration name: (identifier) @name)) @definition.class
        (lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function))) @definition.function
    """,
    import_query="""
        (import_statement) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
        (call_expression function: (member_expression property: (property_identifier) @name)) @call
    """,
    container_types=["class_declaration", "class"],
)

_TYPESCRIPT_SPEC = LanguageSpec(
    extensions=[".ts", ".tsx"],
    symbol_query="""
        (function_declaration name: (identifier) @name) @definition.function
        (class_declaration name: (type_identifier) @name) @definition.class
        (method_definition name: (property_identifier) @name) @definition.method
        (interface_declaration name: (type_identifier) @name) @definition.type
        (type_alias_declaration name: (type_identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function))) @definition.function
    """,
    import_query="""
        (import_statement) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
        (call_expression function: (member_expression property: (property_identifier) @name)) @call
    """,
    container_types=["class_declaration", "interface_declaration"],
)

_GO_SPEC = LanguageSpec(
    extensions=[".go"],
    symbol_query="""
        (function_declaration name: (identifier) @name) @definition.function
        (method_declaration name: (field_identifier) @name) @definition.method
        (type_declaration (type_spec name: (type_identifier) @name)) @definition.type
    """,
    import_query="""
        (import_declaration) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
        (call_expression function: (selector_expression field: (field_identifier) @name)) @call
    """,
    container_types=[],
)

_RUST_SPEC = LanguageSpec(
    extensions=[".rs"],
    symbol_query="""
        (function_item name: (identifier) @name) @definition.function
        (struct_item name: (type_identifier) @name) @definition.class
        (enum_item name: (type_identifier) @name) @definition.type
        (trait_item name: (type_identifier) @name) @definition.type
        (impl_item type: (type_identifier) @name) @definition.class
        (type_item name: (type_identifier) @name) @definition.type
    """,
    import_query="""
        (use_declaration) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
        (call_expression function: (field_expression field: (field_identifier) @name)) @call
    """,
    container_types=["impl_item"],
)

_JAVA_SPEC = LanguageSpec(
    extensions=[".java"],
    symbol_query="""
        (method_declaration name: (identifier) @name) @definition.method
        (class_declaration name: (identifier) @name) @definition.class
        (interface_declaration name: (identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (constructor_declaration name: (identifier) @name) @definition.method
    """,
    import_query="""
        (import_declaration) @import
    """,
    call_query="""
        (method_invocation name: (identifier) @name) @call
    """,
    container_types=["class_declaration", "interface_declaration", "enum_declaration"],
)

_PHP_SPEC = LanguageSpec(
    extensions=[".php"],
    symbol_query="""
        (function_definition name: (name) @name) @definition.function
        (class_declaration name: (name) @name) @definition.class
        (method_declaration name: (name) @name) @definition.method
        (interface_declaration name: (name) @name) @definition.type
        (trait_declaration name: (name) @name) @definition.type
    """,
    import_query="""
        (namespace_use_declaration) @import
    """,
    call_query="""
        (function_call_expression function: (name) @name) @call
        (member_call_expression name: (name) @name) @call
    """,
    container_types=["class_declaration", "interface_declaration", "trait_declaration"],
)

_DART_SPEC = LanguageSpec(
    extensions=[".dart"],
    symbol_query="""
        (function_signature name: (identifier) @name) @definition.function
        (class_definition name: (identifier) @name) @definition.class
        (method_signature name: (identifier) @name) @definition.method
        (enum_declaration name: (identifier) @name) @definition.type
    """,
    import_query="""
        (import_or_export) @import
    """,
    call_query="",
    container_types=["class_definition"],
)

_CSHARP_SPEC = LanguageSpec(
    extensions=[".cs"],
    symbol_query="""
        (method_declaration name: (identifier) @name) @definition.method
        (class_declaration name: (identifier) @name) @definition.class
        (interface_declaration name: (identifier) @name) @definition.type
        (struct_declaration name: (identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (constructor_declaration name: (identifier) @name) @definition.method
    """,
    import_query="""
        (using_directive) @import
    """,
    call_query="""
        (invocation_expression function: (identifier) @name) @call
        (invocation_expression function: (member_access_expression name: (identifier) @name)) @call
    """,
    container_types=["class_declaration", "interface_declaration", "struct_declaration"],
)

_C_SPEC = LanguageSpec(
    extensions=[".c", ".h"],
    symbol_query="""
        (function_definition declarator: (function_declarator declarator: (identifier) @name)) @definition.function
        (struct_specifier name: (type_identifier) @name) @definition.type
        (enum_specifier name: (type_identifier) @name) @definition.type
        (type_definition declarator: (type_identifier) @name) @definition.type
    """,
    import_query="""
        (preproc_include) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
    """,
    container_types=[],
)

_CPP_SPEC = LanguageSpec(
    extensions=[".cpp", ".cc", ".cxx", ".hpp", ".hxx", ".hh"],
    symbol_query="""
        (function_definition declarator: (function_declarator declarator: (identifier) @name)) @definition.function
        (function_definition declarator: (function_declarator declarator: (qualified_identifier name: (identifier) @name))) @definition.function
        (class_specifier name: (type_identifier) @name) @definition.class
        (struct_specifier name: (type_identifier) @name) @definition.type
    """,
    import_query="""
        (preproc_include) @import
    """,
    call_query="""
        (call_expression function: (identifier) @name) @call
        (call_expression function: (field_expression field: (field_identifier) @name)) @call
    """,
    container_types=["class_specifier"],
)

_ELIXIR_SPEC = LanguageSpec(
    extensions=[".ex", ".exs"],
    symbol_query="""
        (call target: (identifier) @_keyword (#any-of? @_keyword "def" "defp" "defmacro") (arguments (identifier) @name)) @definition.function
        (call target: (identifier) @_keyword (#any-of? @_keyword "defmodule") (arguments (alias) @name)) @definition.class
    """,
    import_query="""
        (call target: (identifier) @_keyword (#any-of? @_keyword "import" "alias" "use" "require")) @import
    """,
    call_query="",
    container_types=[],
)

_RUBY_SPEC = LanguageSpec(
    extensions=[".rb", ".rake", ".gemspec"],
    symbol_query="""
        (method name: (identifier) @name) @definition.function
        (singleton_method name: (identifier) @name) @definition.function
        (class name: (constant) @name) @definition.class
        (module name: (constant) @name) @definition.class
    """,
    import_query="""
        (call method: (identifier) @_m (#any-of? @_m "require" "require_relative" "include")) @import
    """,
    call_query="""
        (call method: (identifier) @name) @call
    """,
    container_types=["class", "module"],
)


# ── Registry ─────────────────────────────────────────────────────────────

LANGUAGE_SPECS: dict[str, LanguageSpec] = {
    "python": _PYTHON_SPEC,
    "javascript": _JAVASCRIPT_SPEC,
    "typescript": _TYPESCRIPT_SPEC,
    "go": _GO_SPEC,
    "rust": _RUST_SPEC,
    "java": _JAVA_SPEC,
    "php": _PHP_SPEC,
    "dart": _DART_SPEC,
    "c_sharp": _CSHARP_SPEC,
    "c": _C_SPEC,
    "cpp": _CPP_SPEC,
    "elixir": _ELIXIR_SPEC,
    "ruby": _RUBY_SPEC,
}


_EXTENSIONS_MAP: dict[str, str] | None = None


def get_extensions_map() -> dict[str, str]:
    """Return cached map from file extension to language name."""
    global _EXTENSIONS_MAP
    if _EXTENSIONS_MAP is None:
        ext_map: dict[str, str] = {}
        for lang_name, spec in LANGUAGE_SPECS.items():
            for ext in spec.extensions:
                ext_map[ext] = lang_name
        _EXTENSIONS_MAP = ext_map
    return _EXTENSIONS_MAP


def detect_language(file_path: str) -> str | None:
    """Detect language from file extension."""
    suffix = Path(file_path).suffix.lower()
    return get_extensions_map().get(suffix)


def get_parser_for_language(lang: str) -> Any:
    """Get a cached tree-sitter Parser for a language.

    Returns None if tree-sitter-language-pack is not installed
    or the language is not supported.
    """
    if lang in _parser_cache:
        return _parser_cache[lang]

    try:
        from tree_sitter_language_pack import get_parser

        parser = get_parser(cast(Any, lang))
        _parser_cache[lang] = parser
        return parser
    except ImportError:
        logger.warning("tree-sitter-language-pack not installed")
        return None
    except Exception as e:
        logger.warning(f"Failed to get parser for {lang}: {e}")
        return None


def get_language_obj(lang: str) -> Any:
    """Get tree-sitter Language object for running queries."""
    try:
        from tree_sitter_language_pack import get_language

        return get_language(cast(Any, lang))
    except ImportError:
        return None
    except Exception as e:
        logger.warning(f"Failed to get language object for {lang}: {e}")
        return None

File: rust/gcode/src/index/languages.rs
//! Language registry with tree-sitter query definitions.
//! Ports 16 language specs from src/gobby/code_index/languages.py.

use std::collections::HashSet;
use tree_sitter::Language;

/// Specification for a single language's tree-sitter queries.
pub struct LanguageSpec {
    pub extensions: &'static [&'static str],
    pub symbol_query: &'static str,
    pub import_query: &'static str,
    pub call_query: &'static str,
    pub container_types: &'static [&'static str],
}

// ── Query Definitions ──────────────────────────────────────────────────

const PYTHON: LanguageSpec = LanguageSpec {
    extensions: &[".py", ".pyi"],
    symbol_query: r#"
        (function_definition name: (identifier) @name) @definition.function
        (class_definition name: (identifier) @name) @definition.class
    "#,
    import_query: r#"
        (import_statement) @import
        (import_from_statement) @import
    "#,
    call_query: r#"
        (call function: (identifier) @name) @call
        (call function: (attribute attribute: (identifier) @name)) @call
    "#,
    container_types: &["class_definition"],
};

const JAVASCRIPT: LanguageSpec = LanguageSpec {
    extensions: &[".js", ".jsx", ".mjs", ".cjs"],
    symbol_query: r#"
        (function_declaration name: (identifier) @name) @definition.function
        (class_declaration name: (identifier) @name) @definition.class
        (method_definition name: (property_identifier) @name) @definition.method
        (export_statement declaration: (function_declaration name: (identifier) @name)) @definition.function
        (export_statement declaration: (class_declaration name: (identifier) @name)) @definition.class
        (lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function))) @definition.function
    "#,
    import_query: r#"
        (import_statement) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
        (call_expression function: (member_expression property: (property_identifier) @name)) @call
    "#,
    container_types: &["class_declaration", "class"],
};

const TYPESCRIPT: LanguageSpec = LanguageSpec {
    extensions: &[".ts", ".tsx"],
    symbol_query: r#"
        (function_declaration name: (identifier) @name) @definition.function
        (class_declaration name: (type_identifier) @name) @definition.class
        (method_definition name: (property_identifier) @name) @definition.method
        (interface_declaration name: (type_identifier) @name) @definition.type
        (type_alias_declaration name: (type_identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function))) @definition.function
        (export_statement declaration: (function_declaration name: (identifier) @name)) @definition.function
        (export_statement declaration: (class_declaration name: (type_identifier) @name)) @definition.class
        (export_statement declaration: (interface_declaration name: (type_identifier) @name)) @definition.type
        (export_statement declaration: (type_alias_declaration name: (type_identifier) @name)) @definition.type
        (export_statement declaration: (enum_declaration name: (identifier) @name)) @definition.type
        (export_statement declaration: (lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function)))) @definition.function
    "#,
    import_query: r#"
        (import_statement) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
        (call_expression function: (member_expression property: (property_identifier) @name)) @call
    "#,
    container_types: &["class_declaration", "interface_declaration"],
};

const GO: LanguageSpec = LanguageSpec {
    extensions: &[".go"],
    symbol_query: r#"
        (function_declaration name: (identifier) @name) @definition.function
        (method_declaration name: (field_identifier) @name) @definition.method
        (type_declaration (type_spec name: (type_identifier) @name)) @definition.type
    "#,
    import_query: r#"
        (import_declaration) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
        (call_expression function: (selector_expression field: (field_identifier) @name)) @call
    "#,
    container_types: &[],
};

const RUST: LanguageSpec = LanguageSpec {
    extensions: &[".rs"],
    symbol_query: r#"
        (function_item name: (identifier) @name) @definition.function
        (struct_item name: (type_identifier) @name) @definition.class
        (enum_item name: (type_identifier) @name) @definition.type
        (trait_item name: (type_identifier) @name) @definition.type
        (impl_item type: (type_identifier) @name) @definition.class
        (type_item name: (type_identifier) @name) @definition.type
    "#,
    import_query: r#"
        (use_declaration) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
        (call_expression function: (field_expression field: (field_identifier) @name)) @call
    "#,
    container_types: &["impl_item"],
};

const JAVA: LanguageSpec = LanguageSpec {
    extensions: &[".java"],
    symbol_query: r#"
        (method_declaration name: (identifier) @name) @definition.method
        (class_declaration name: (identifier) @name) @definition.class
        (interface_declaration name: (identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (constructor_declaration name: (identifier) @name) @definition.method
    "#,
    import_query: r#"
        (import_declaration) @import
    "#,
    call_query: r#"
        (method_invocation name: (identifier) @name) @call
    "#,
    container_types: &["class_declaration", "interface_declaration", "enum_declaration"],
};

const PHP: LanguageSpec = LanguageSpec {
    extensions: &[".php"],
    symbol_query: r#"
        (function_definition name: (name) @name) @definition.function
        (class_declaration name: (name) @name) @definition.class
        (method_declaration name: (name) @name) @definition.method
        (interface_declaration name: (name) @name) @definition.type
        (trait_declaration name: (name) @name) @definition.type
    "#,
    import_query: r#"
        (namespace_use_declaration) @import
    "#,
    call_query: r#"
        (function_call_expression function: (name) @name) @call
        (member_call_expression name: (name) @name) @call
    "#,
    container_types: &["class_declaration", "interface_declaration", "trait_declaration"],
};

const DART: LanguageSpec = LanguageSpec {
    extensions: &[".dart"],
    symbol_query: r#"
        (function_signature name: (identifier) @name) @definition.function
        (class_definition name: (identifier) @name) @definition.class
        (method_signature (function_signature (identifier) @name)) @definition.method
        (enum_declaration name: (identifier) @name) @definition.type
    "#,
    import_query: r#"
        (import_or_export) @import
    "#,
    call_query: "",
    container_types: &["class_definition"],
};

const CSHARP: LanguageSpec = LanguageSpec {
    extensions: &[".cs"],
    symbol_query: r#"
        (method_declaration name: (identifier) @name) @definition.method
        (class_declaration name: (identifier) @name) @definition.class
        (interface_declaration name: (identifier) @name) @definition.type
        (struct_declaration name: (identifier) @name) @definition.type
        (enum_declaration name: (identifier) @name) @definition.type
        (constructor_declaration name: (identifier) @name) @definition.method
    "#,
    import_query: r#"
        (using_directive) @import
    "#,
    call_query: r#"
        (invocation_expression function: (identifier) @name) @call
        (invocation_expression function: (member_access_expression name: (identifier) @name)) @call
    "#,
    container_types: &["class_declaration", "interface_declaration", "struct_declaration"],
};

const C_LANG: LanguageSpec = LanguageSpec {
    extensions: &[".c", ".h"],
    symbol_query: r#"
        (function_definition declarator: (function_declarator declarator: (identifier) @name)) @definition.function
        (struct_specifier name: (type_identifier) @name) @definition.type
        (enum_specifier name: (type_identifier) @name) @definition.type
        (type_definition declarator: (type_identifier) @name) @definition.type
    "#,
    import_query: r#"
        (preproc_include) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
    "#,
    container_types: &[],
};

const CPP: LanguageSpec = LanguageSpec {
    extensions: &[".cpp", ".cc", ".cxx", ".hpp", ".hxx", ".hh"],
    symbol_query: r#"
        (function_definition declarator: (function_declarator declarator: (identifier) @name)) @definition.function
        (function_definition declarator: (function_declarator declarator: (qualified_identifier name: (identifier) @name))) @definition.function
        (class_specifier name: (type_identifier) @name) @definition.class
        (struct_specifier name: (type_identifier) @name) @definition.type
    "#,
    import_query: r#"
        (preproc_include) @import
    "#,
    call_query: r#"
        (call_expression function: (identifier) @name) @call
        (call_expression function: (field_expression field: (field_identifier) @name)) @call
    "#,
    container_types: &["class_specifier"],
};

const ELIXIR: LanguageSpec = LanguageSpec {
    extensions: &[".ex", ".exs"],
    symbol_query: r#"
        (call target: (identifier) @_keyword (#any-of? @_keyword "def" "defp" "defmacro") (arguments (identifier) @name)) @definition.function
        (call target: (identifier) @_keyword (#any-of? @_keyword "defmodule") (arguments (alias) @name)) @definition.class
    "#,
    import_query: r#"
        (call target: (identifier) @_keyword (#any-of? @_keyword "import" "alias" "use" "require")) @import
    "#,
    call_query: "",
    container_types: &[],
};

const RUBY: LanguageSpec = LanguageSpec {
    extensions: &[".rb", ".rake", ".gemspec"],
    symbol_query: r#"
        (method name: (identifier) @name) @definition.function
        (singleton_method name: (identifier) @name) @definition.function
        (class name: (constant) @name) @definition.class
        (module name: (constant) @name) @definition.class
    "#,
    import_query: r#"
        (call method: (identifier) @_m (#any-of? @_m "require" "require_relative" "include")) @import
    "#,
    call_query: r#"
        (call method: (identifier) @name) @call
    "#,
    container_types: &["class", "module"],
};

const MARKDOWN: LanguageSpec = LanguageSpec {
    extensions: &[".md", ".markdown"],
    symbol_query: r#"
        (atx_heading heading_content: (_) @name) @definition.section
        (setext_heading heading_content: (_) @name) @definition.section
    "#,
    import_query: "",
    call_query: "",
    container_types: &[],
};

const YAML: LanguageSpec = LanguageSpec {
    extensions: &[".yaml", ".yml"],
    symbol_query: r#"
        (block_mapping_pair key: (_) @name) @definition.property
    "#,
    import_query: "",
    call_query: "",
    container_types: &["block_mapping_pair"],
};

const JSON_LANG: LanguageSpec = LanguageSpec {
    extensions: &[".json", ".jsonc"],
    symbol_query: r#"
        (pair key: (string (string_content) @name)) @definition.property
    "#,
    import_query: "",
    call_query: "",
    container_types: &["pair"],
};

// ── Registry ───────────────────────────────────────────────────────────

/// All supported languages and their specs.
const SPECS: &[(&str, &LanguageSpec)] = &[
    ("python", &PYTHON),
    ("javascript", &JAVASCRIPT),
    ("typescript", &TYPESCRIPT),
    ("go", &GO),
    ("rust", &RUST),
    ("java", &JAVA),
    ("php", &PHP),
    ("dart", &DART),
    ("csharp", &CSHARP),
    ("c", &C_LANG),
    ("cpp", &CPP),
    ("elixir", &ELIXIR),
    ("ruby", &RUBY),
    ("markdown", &MARKDOWN),
    ("yaml", &YAML),
    ("json", &JSON_LANG),
];

/// Detect language name from file extension.
pub fn detect_language(file_path: &str) -> Option<&'static str> {
    let path = std::path::Path::new(file_path);
    let ext = path
        .extension()
        .map(|e| format!(".{}", e.to_string_lossy().to_lowercase()))?;

    for (name, spec) in SPECS {
        if spec.extensions.contains(&ext.as_str()) {
            return Some(name);
        }
    }
    None
}

/// Get the language spec for a given language name.
pub fn get_spec(lang: &str) -> Option<&'static LanguageSpec> {
    SPECS.iter().find(|(name, _)| *name == lang).map(|(_, s)| *s)
}

/// Get the tree-sitter Language object for a given language name.
pub fn get_ts_language(lang: &str) -> Option<Language> {
    let lang_fn = match lang {
        "python" => tree_sitter_python::LANGUAGE,
        "javascript" => tree_sitter_javascript::LANGUAGE,
        "typescript" => tree_sitter_typescript::LANGUAGE_TYPESCRIPT,
        "go" => tree_sitter_go::LANGUAGE,
        "rust" => tree_sitter_rust::LANGUAGE,
        "java" => tree_sitter_java::LANGUAGE,
        "c" => tree_sitter_c::LANGUAGE,
        "cpp" => tree_sitter_cpp::LANGUAGE,
        "csharp" => tree_sitter_c_sharp::LANGUAGE,
        "ruby" => tree_sitter_ruby::LANGUAGE,
        "php" => tree_sitter_php::LANGUAGE_PHP,
        "swift" => tree_sitter_swift::LANGUAGE,
        // kotlin 0.3 is built against tree-sitter 0.20, incompatible with 0.24
        // TODO: update when tree-sitter-kotlin publishes a 0.24-compatible release
        "kotlin" => return None,
        "dart" => tree_sitter_dart::LANGUAGE,
        "elixir" => tree_sitter_elixir::LANGUAGE,
        "json" => tree_sitter_json::LANGUAGE,
        "yaml" => tree_sitter_yaml::LANGUAGE,
        "markdown" => tree_sitter_md::LANGUAGE,
        _ => return None,
    };
    Some(lang_fn.into())
}

/// Set of all file extensions supported for AST parsing.
pub fn supported_extensions() -> HashSet<&'static str> {
    let mut exts = HashSet::new();
    for (_, spec) in SPECS {
        for ext in spec.extensions {
            exts.insert(*ext);
        }
    }
    exts
}
